README.md
# PAL Locomotion Tutorials

This package is intended to serve as a tutorial with worked examples and utilities for a better understanding of the DCM controller implemented at PAL Robotics.

This package includes:
* The [balance_action](./src/balance_action.cpp): an implementation of an action of the DCM controller to control the Center of Mass (CoM) of the robot making it describe a sinusoidal wave.
* The [static_steps_command](./src/static_steps_command.cpp): is a description of the different interfaces available to send footsteps to the robot when performing quasi-static walking.
* The [static_steps_execution_node](./src/static_steps_execution_node.cpp): that creates a list of footsteps to command the robot to move on a straight line.

## Balance action

The balance action is a specific action from the state machine inside the DCM controller. The [talos_pal_locomotion](https://gitlab/control/talos_pal_locomotion) package contains a brief explanation of the state machine and the available actions inside the DCM controller.

The balance action shows the capabilities of the robot to maintain balance while moving its CoM along each coordinate axis. The implementation is based on the paper [*Bipedal walking control based on Capture Point dynamics*][bipedal_walking_control_link], by Englsberger, Ott, Roa, Albu-Schäfer and Hirzinger. Within the balance action node, the CoM of the robot is commanded to move on a sinusoidal trajectory of a user-provided frequency and amplitude. The robot maintains balance while performing the motion.

### Code of Balance Action

* The [balance_action.cpp](./src/balance_action.cpp) file is an implementation of the action by controlling the dynamics of the [paper cited above][bipedal_walking_control_link] in a plugin, that uses the information from the Biped controller.
* The [pal_locomotion_tutorials_plugins.cpp](./src/pal_locomotion_tutorials_plugins.cpp) file uses the plugin library to export the class from the package.
* The [balance_action_node.cpp](./src/balance_action_node.cpp) file is an implementation of a node that pushes the created plugin into the `biped_walking_dcm_controller` and starts the action.

#### Execution of Balance Action

* In simulation
    1. To launch the action, first launch the simulation:
        ```
        roslaunch talos_gazebo talos_gazebo.launch
        ```
    2. then launch the DCM controller to start the biped controller:
        ```
        roslaunch talos_pal_locomotion talos_dcm_walking_controller.launch estimator:=kinematic_estimator_params
        ```
* In the real robot

    To launch the demo on the **real robot**, 

    1. Deploy this package on the robot, and then reboot the robot
    2. Launch the default controllers on the robot, once the robot is in walking pose, place the robot on ground
        ```
        roslaunch talos_controller_configuration default_controllers.launch
        ```
    3. Kill the default controllers
    4. Launch the DCM Controller on the robot (refer to the handbook for the command)

Now, To launch the balance action node, run the following command:
```
rosrun pal_locomotion_tutorials balance_action_node
```

<span style="color:red">**Caution: When running in the robot, increase the limits in small fractions to avoid breaking the robot (or) to avoid completely destabilize the robot** </span>

Open `rqt_reconfigure` and then navigate to the following option `biped_walking_dcm_controller`->`balance control_params` as shown in the figure:

![Balance Control Parameters](./images/balance_action_rqt_reconf.png#center)

With the parameters in the `rqt_reconfigure`, set the desired frequency and amplitude of the desired sinusoidal motion in the desired direction. The below picture shows a glimpse of robot in the simulation performing the motion, while maintaining the balance. 

![TALOS Squat](./images/balance_action_talos_squat.png#center)

## Execution of Static Steps

The following nodes are an example of how to send footsteps to the robot by using an action, a service, a publisher or a library.

- [static_steps_command](./src/static_steps_command.cpp)
- [static_steps_execution_node](./src/static_steps_execution_node.cpp)

The first one uses the three ROS interfaces to send a list o footsteps (i.e action, service, publisher). In that example the footsteps are defined either as velocity commands or as a poses in the world. It also contains an example of how to define trajectories in Cartesian space for the swing leg.

![TALOS Static Steps Execution](./images/static_steps_execution_talos.png#center)

The second one uses the `StaticStepExecutor` library from pal_locomotion. The main difference with respect to the first node is that the previous one uses the `static_walk` action from [pal_locomotion_actions](https://gitlab/control/pal_locomotion_actions), while this node uses the `stand_one_leg_action, stand_down_leg_action` actions. In these actions there is no tolerance in the CoM considered to go to the next state, it is done by time.

**Parameters:**
- The parameter `d` (or) `ds_duration` is the transition duration of each commanded step.
- The parameter `u` (or) `swing_duration` is the swing duration of each commanded step.
- The parameter `s` (or) `step_num` is the number of steps to be executed.
- The parameter `x` (or) `x_dist` is footstep distance in the x direction.
- The parameter `y` (or) `foot_separation` is footstep distance in the y direction.

For example, to perform 5 steps (with the suggested parameters by the static_steps_execution_node):

* In simulation
    1. To launch the action, first launch the simulation:
        ```
        roslaunch talos_gazebo talos_gazebo.launch
        ```
    2. then launch the DCM controller to start the biped controller:
        ```
        roslaunch talos_pal_locomotion talos_dcm_walking_controller.launch estimator:=kinematic_estimator_params
        ```
    3. and run the static_steps_execution_node
        ```
        rosrun pal_locomotion_tutorials static_steps_execution_node -d 2.0 -u 2.0 -s 5 -x 0.15 -y 0.1
        ```

* In the robot
    1. Follow the procedure expained above to execute the balance action, and run
        ```
        rosrun pal_locomotion_tutorials static_steps_execution_node -d 2.0 -u 2.0 -s 5 -x 0.15 -y 0.1
        ```

[bipedal_walking_control_link]: https://ieeexplore.ieee.org/document/6094435


CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package pal_training_actions
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

0.1.7 (2020-07-28)
------------------
* Merge branch 'add_ex_doc' into 'master'
  Added example to execute static steps
  See merge request control/pal_locomotion_tutorials!15
* Added example to execute static steps
* Merge branch 'imgs' into 'master'
  Imgs
  See merge request control/pal_locomotion_tutorials!14
* Center imgs
* Reduce imgs
* Merge branch 'documentation_review' into 'master'
  Documentation review
  See merge request control/pal_locomotion_tutorials!13
* Documentation review
* Update README
* Contributors: Adria Roig, narcismiguel

0.1.6 (2020-07-08)
------------------
* Merge branch 'talos_6_tests' into 'master'
  Modify step distances according to talos 6 tests
  See merge request control/pal_locomotion_tutorials!12
* Modify step distances according to talos 6 tests
* Contributors: Adria Roig, victor

0.1.5 (2020-05-15)
------------------
* Merge branch 'swing_interp' into 'master'
  Change API to swing trajectory interpolator
  See merge request control/pal_locomotion_tutorials!11
* Change API to swing trajectory interpolator
* Contributors: Adrià Roig, victor

0.1.4 (2020-03-18)
------------------
* Merge branch 'static_step_refactor' into 'master'
  added changes of the static step executor
  See merge request control/pal_locomotion_tutorials!10
* added changes of the static step executor
* Contributors: Sai Kishor Kothakota, victor

0.1.3 (2020-03-10)
------------------
* Merge branch 'walking_apu' into 'master'
  Walking apu
  See merge request control/pal_locomotion_tutorials!9
* Fix typo
* Fix typos
* Fixed wrong rebase
* Add node showing the walking API
* Contributors: Adria Roig, victor

0.1.2 (2020-03-05)
------------------
* Merge branch 'rename_pkg' into 'master'
  Rename package to pal_locomotion_tutorials
  See merge request control/pal_training_actions!8
* Rename package to pal_locomotion_tutorials
* Contributors: Sai Kishor Kothakota, victor

0.1.1 (2020-03-05)
------------------
* Merge branch 'static_step_executor' into 'master'
  use static_step_executor from pal_locomotion
  See merge request control/pal_training_actions!7
* use static_step_executor from pal_locomotion
* Contributors: Sai Kishor Kothakota, victor

0.1.0 (2020-03-03)
------------------
* Merge branch 'static_walking' into 'master'
  Static walking
  See merge request control/pal_training_actions!5
* Merge branch 'update_amplitudes' into 'master'
  Update max amplitudes along y and z
  See merge request control/pal_training_actions!6
* Update max amplitudes along y and z
* Add more options to static steps execution node
* Experiments  of static walking on the robot
* update README
* removed empty files
* added option to choose the number of steps at execution time
* Add static steps node
* Merge branch 'readme' into 'master'
  Readme
  See merge request control/pal_training_actions!4
* added README on balance action
* removed unsued centre_foot_point computation
* Contributors: Adria Roig, Sai Kishor Kothakota, saikishor, victor

0.0.2 (2019-12-09)
------------------
* Merge branch 'staging_fixes' into 'master'
  Updated the service message API as of staging
  See merge request control/pal_training_actions!3
* Updated the service message API as of staging
* Contributors: Sai Kishor Kothakota

0.0.1 (2019-12-09)
------------------
* Merge branch 'stable_changes' into 'master'
  Stable changes
  See merge request adriaroig/pal_training_actions!2
* add missing install files
* limit the balance action parameters
* Updated the service message API as of stable
* removed unused icp_walking_sim dependency
* Initial commit
* Contributors: Adria Roig, Sai Kishor Kothakota


src/balance_action.cpp
/*
 * Copyright 2019 PAL Robotics SL. All Rights Reserved
 *
 * Unauthorized copying of this file, via any medium is strictly prohibited,
 * unless it was supplied under the terms of a license agreement or
 * nondisclosure agreement with PAL Robotics SL. In this case it may not be
 * copied or disclosed except in accordance with the terms of that agreement.
 */

#include <pal_locomotion_tutorials/balance_action.h>
#include <math_utils/geometry_tools.h>
#include <ros/ros.h>
#include <ddynamic_reconfigure/ddynamic_reconfigure.h>

using namespace pal_locomotion;
using namespace math_utils;

BalanceAction::BalanceAction(ros::NodeHandle &nh, BController *bController)
  :nh_(nh), time_(0.0)
{
  if (!configure(nh, bController, property_bag::PropertyBag()))
  {
    PAL_THROW_DEFAULT("problem configuring actoin");
  }
}

BalanceAction::~BalanceAction()
{
}

bool BalanceAction::configure(ros::NodeHandle &nh, BController *bController,
                              const property_bag::PropertyBag &parameters)
{
  bc_ = bController;
  ddr_.reset(new ddynamic_reconfigure::DDynamicReconfigure(
      ros::NodeHandle(nh, "balance_control_params")));
  ddr_->RegisterVariable(&params_.sin_freq_z_, "frequency_z", 0.0, 2.0);
  ddr_->RegisterVariable(&params_.sin_amp_z_, "amplitude_z", 0.0, 0.08);
  ddr_->RegisterVariable(&params_.sin_freq_y_, "frequency_y", 0.0, 2.0);
  ddr_->RegisterVariable(&params_.sin_amp_y_, "amplitude_y", 0.0, 0.085);
  ddr_->RegisterVariable(&params_.sin_freq_x_, "frequency_x", 0.0, 2.0);
  ddr_->RegisterVariable(&params_.sin_amp_x_, "amplitude_x", 0.0, 0.05);
  ddr_->RegisterVariable(&params_.Kp_x_, "Kp_x", 0.0, 30.0);
  ddr_->RegisterVariable(&params_.Kd_x_, "kd_x", 0.0, 10.0);
  ddr_->RegisterVariable(&params_.Kp_y_, "Kp_y", 0.0, 30.0);
  ddr_->RegisterVariable(&params_.Kd_y_, "kd_y", 0.0, 10.0);
  ddr_->RegisterVariable(&params_.Kp_z_, "Kp_z", 0.0, 30.0);
  ddr_->RegisterVariable(&params_.Kd_z_, "kd_z", 0.0, 10.0);
  ddr_->publishServicesTopics();

  // For realtime publisher
  n_com_states_ = 6;
  com_states_pub_.reset(
    new realtime_tools::RealtimePublisher<std_msgs::Float64MultiArray>(nh, "com_states", 1));

  desired_motions_buffer_.writeFromNonRT(default_motion_command_);
  sub_command_ = nh.subscribe<std_msgs::Float64MultiArray>("motionCommand", 1, &BalanceAction::commandCallback, this);

  return true;
}

bool BalanceAction::enterHook(const ros::Time &time)
{
  std::vector<Side> stance_id;
  stance_id.push_back(+Side::LEFT);
  stance_id.push_back(+Side::RIGHT);
  std::vector<Side> swingLegIds;
  bc_->setStanceLegIDs(stance_id);
  bc_->setSwingLegIDs(swingLegIds);

  bc_->setWeightDistribution(0.5);
  bc_->setActualSupportType(+SupporType::DS);

  eMatrixHom actualLeftFootPose = bc_->getActualFootPose(+Side::LEFT);
  eMatrixHom actualRightFootPose = bc_->getActualFootPose(+Side::RIGHT);

  ini_target_ = (actualLeftFootPose.translation() + actualRightFootPose.translation()) / 2.;

  return true;
}

bool BalanceAction::cycleHook(const ros::Time &time)
{
  // Get the actual pose of the foot, so that we can find the center point of the foot
  eMatrixHom act_left_foot_pose = bc_->getActualFootPose(+Side::LEFT);
  eMatrixHom act_right_foot_pose = bc_->getActualFootPose(+Side::RIGHT);

  // Create the local coordinate frame of reference
  eMatrixHom local_coordinate_frame =
      interpolateBetweenTransforms(act_left_foot_pose, act_right_foot_pose);

  eMatrixHom2d coordinate_system_2d;
  pal::convert(local_coordinate_frame, coordinate_system_2d);

  std::vector<double>& desired_motionCommand = *desired_motions_buffer_.readFromRT();

  Eigen::Vector3d desired_pos;
  Eigen::Vector3d desired_vel;
  Eigen::Vector3d desired_acc;

  desired_pos.x() = params_.sin_amp_x_*sin(2*M_PI* time_*params_.sin_freq_x_);
  // desired_pos.y() = params_.sin_amp_y_*sin(2*M_PI* time_*params_.sin_freq_y_);
  desired_pos.y() = desired_motionCommand[2]*sin(2*M_PI* desired_motionCommand[3]);
  desired_pos.z() = params_.sin_amp_z_*sin(2*M_PI* time_*params_.sin_freq_z_) +  bc_->getParameters()->z_height_;

  desired_vel.x() = params_.sin_amp_x_*2*M_PI*params_.sin_freq_x_*cos(2*M_PI* time_*params_.sin_freq_x_);
  // desired_vel.y() = params_.sin_amp_y_*2*M_PI*params_.sin_freq_y_*cos(2*M_PI* time_*params_.sin_freq_y_);
  desired_vel.y() = desired_motionCommand[2]*2*M_PI*desired_motionCommand[3]*cos(2*M_PI* time_*desired_motionCommand[3]);
  desired_vel.z() = params_.sin_amp_z_*2*M_PI*params_.sin_freq_z_*cos(2*M_PI* time_*params_.sin_freq_z_);

  desired_acc.x() = -params_.sin_amp_x_*pow(2*M_PI*params_.sin_freq_x_, 2)*sin(2*M_PI* time_*params_.sin_freq_x_);
  // desired_acc.y() = -params_.sin_amp_y_*pow(2*M_PI*params_.sin_freq_y_, 2)*sin(2*M_PI* time_*params_.sin_freq_y_);
  desired_acc.y() = -desired_motionCommand[2]*pow(2*M_PI*desired_motionCommand[3], 2)*sin(2*M_PI* time_*desired_motionCommand[3]);
  desired_acc.z() = -params_.sin_amp_z_*pow(2*M_PI*params_.sin_freq_z_, 2)*sin(2*M_PI* time_*params_.sin_freq_z_);

  if (com_states_pub_ && com_states_pub_->trylock())
  {   
      current_com_pos_ = bc_->getActualCOMPosition();
      com_states_pub_->msg_.data.resize(n_com_states_);
      com_states_pub_->msg_.data[0] = current_com_pos_.x();
      com_states_pub_->msg_.data[1] = current_com_pos_.y();
      com_states_pub_->msg_.data[2] = current_com_pos_.z();
      com_states_pub_->msg_.data[3] = desired_pos.x();
      com_states_pub_->msg_.data[4] = desired_pos.y();
      com_states_pub_->msg_.data[5] = desired_pos.z();
      com_states_pub_->unlockAndPublish();
  }
  // Our actual COP and desired Pcmp are at the origin of the coordinate system as we just
  // wanted to balance the system mainting them at that point
  // Eigen::Vector2d act_COP = coordinate_system_2d.translation();
  // Eigen::Vector2d desiredPcmp = coordinate_system_2d.translation() + Eigen::Vector2d(dx, dy);
  // Eigen::Vector2d desiredPcmp_vel = Eigen::Vector2d::Zero();
  // desire_z = bc_->getParameters()->z_height_ + desire_z;

  // double omega = sqrt(bc_->getParameters()->gravity_ / z);
  Eigen::Vector2d actCOM = bc_->getActualCOMPosition2d();
  Eigen::Vector2d actCOMvel = bc_->getActualCOMVelocity2d();

  // Compute the ICP & Pcmp and next operations
  // Eigen::Vector2d actICP =
  //     this->computeActICP(omega, actCOM, actCOMvel, coordinate_system_2d.translation());
  // Eigen::Vector2d Pcmp = this->computePcmp(bc_->getParameters()->icp_gain_, omega, actICP, desiredPcmp, desiredPcmp_vel);

  // Here we do work in local cooridnates for accel and later converert it to global
  // for the computation for desired COM_vel and COM, we use actual ones as the for next control loop the desired is expected to be actual
  double dt = bc_->getControllerDt().toSec();
  current_com_pos_ = bc_->getActualCOMPosition();
  
  desired_acc.x() = desired_acc.x() + params_.Kp_x_* (desired_pos.x() - current_com_pos_.x()) + params_.Kd_x_*(desired_vel.x() - bc_->getActualCOMVelocity().x());
  desired_acc.y() = desired_acc.y() + params_.Kp_y_* (desired_pos.y() - current_com_pos_.y()) + params_.Kp_y_*(desired_vel.y() - bc_->getActualCOMVelocity().y());
  desired_acc.z() = desired_acc.z() + params_.Kp_z_* (desired_pos.z() - current_com_pos_.z()) + params_.Kp_z_*(desired_vel.z() - bc_->getActualCOMVelocity().z());

  // Eigen::Vector2d desiredCOM_acc = params_.Kp_*(desiredCOM - bc_->getActualCOMPosition()) + params_.Kd_*(desiredCOM_vel - bc_->getActualCOMVelocity()) + 
  //                                  omega * omega* ((actCOM - coordinate_system_2d.translation()) - (Pcmp-coordinate_system_2d.translation()));
  // Eigen::Vector2d desiredCOM_vel = actCOMvel + (desiredCOM_acc *dt);
  // Eigen::Vector2d desiredCOM = actCOM + (desiredCOM_vel * dt) + (0.5 * desiredCOM_acc * dt * dt);

  // Set the desired COM attributes to the biped controller
  bc_->setDesiredCOMPosition(eVector3(desired_pos.x(), desired_pos.y(), desired_pos.z())); // It's only for debugging
  bc_->setDesiredCOMVelocity(eVector3(desired_vel.x(), desired_vel.y(), desired_vel.z())); // It's only for debugging
  bc_->setDesiredCOMAcceleration(eVector3(desired_acc.x(), desired_acc.y(), desired_acc.z())); // This is for the control point

  // Set the ICP and COP attributes to the biped controller
  // All the below are only for debugging, but not used for the control point of view.
  // bc_->setDesiredICP(eVector3(desiredPcmp.x(), desiredPcmp.y(), 0.));
  // bc_->setDesiredCOPReference(eVector3(act_COP.x(), act_COP.y(), 0.));
  // bc_->setDesiredCOPComputed(eVector3(Pcmp.x(), Pcmp.y(), 0.0));
  // bc_->setActualICP(eVector3(actICP.x(), actICP.y(), 0.0));
  time_ += dt;
  return true;
}

bool BalanceAction::isOverHook(const ros::Time &time)
{
  if (bc_->getStateMachine()->queue_size() > 1)
  {
    return true;
  }
  return false;
}

bool BalanceAction::endHook(const ros::Time &time)
{
  ROS_INFO_STREAM("Balance ds end hook, time: " << time.toSec());
  return true;
}


eVector2 BalanceAction::computePcmp(double K, double omega, const eVector2 &icp_act,
                                    const eVector2 &icp_des, const eVector2 &icp_des_vel)
{
  eVector2 Pcmp = icp_act - ((1 / omega) * icp_des_vel) + (K * (icp_act - icp_des));
  return Pcmp;
}

Eigen::Vector2d BalanceAction::computeActICP(double omega, const Eigen::Vector2d &com,
                                             const Eigen::Vector2d &com_d,
                                             const Eigen::Vector2d &local_coord)
{
  Eigen::Vector2d icp = (com - local_coord) + (com_d / omega);
  return (icp + local_coord);
}

void BalanceAction::commandCallback(const std_msgs::Float64MultiArrayConstPtr& msg){
  desired_motions_buffer_.writeFromNonRT(msg->data);
}

src/balance_action_node.cpp
/*
 * Copyright 2019 PAL Robotics SL. All Rights Reserved
 *
 * Unauthorized copying of this file, via any medium is strictly prohibited,
 * unless it was supplied under the terms of a license agreement or
 * nondisclosure agreement with PAL Robotics SL. In this case it may not be
 * copied or disclosed except in accordance with the terms of that agreement.
 */

#include <pal_locomotion_msgs/ActionWithParameters.h>
#include <pal_locomotion_msgs/PushActions.h>
#include <boost/archive/text_oarchive.hpp>
#include <boost/archive/text_iarchive.hpp>
#include <property_bag/serialization/property_bag_boost_serialization.h>
#include <ros/ros.h>

int main(int argc, char **argv)
{
  // Set up ROS.
  ros::init(argc, argv, "balance_action_training");
  ros::NodeHandle nh;

  ros::ServiceClient client = nh.serviceClient<pal_locomotion_msgs::PushActions>(
      "/biped_walking_dcm_controller/push_actions");

  if (!client.waitForExistence(ros::Duration(5.0)))
  {
    ROS_ERROR_STREAM("Cannot connect to the service : " << client.getService());
    return 1;
  }

  pal_locomotion_msgs::PushActions push_actions_request;

  pal_locomotion_msgs::ActionWithParameters new_action;
  new_action.action_type = "pal_locomotion::BalanceAction";
  property_bag::PropertyBag parameters;

  std::stringstream ss;
  boost::archive::text_oarchive oa(ss);
  oa << parameters;
  new_action.action_parameters = ss.str();
  push_actions_request.request.actions.push_back(new_action);

  if (client.call(push_actions_request))
  {
    ROS_INFO("Succesfully called service");
  }
  else
  {
    ROS_ERROR("Failed to call service");
    return 1;
  }
}


include/pal_locomotion_tutorials/balance_action.h
/*
 * Copyright 2019 PAL Robotics SL. All Rights Reserved
 *
 * Unauthorized copying of this file, via any medium is strictly prohibited,
 * unless it was supplied under the terms of a license agreement or
 * nondisclosure agreement with PAL Robotics SL. In this case it may not be
 * copied or disclosed except in accordance with the terms of that agreement.
 */
#ifndef BALANCING_ACTION_H
#define BALANCING_ACTION_H

#include <pal_locomotion/biped_controller.h>
#include <pal_locomotion/state_machine/walking_action_base.h>
// For realtime publisher and subscriber
#include <realtime_tools/realtime_publisher.h>
#include <realtime_tools/realtime_buffer.h>
#include <std_msgs/Float64MultiArray.h>
#include <ros/node_handle.h>
#include<cmath>
//#include <icp_walking_sim/icp_walking.h>

namespace pal_locomotion
{
struct BalanceActionParameters : public ariles::ConfigurableBase
{
  BalanceActionParameters()
  {
    setDefaults();
  }

  void setDefaults()
  {
    sin_freq_z_ = 1.0;
    sin_amp_z_ = 0.0;
    sin_freq_y_ = 1.0;
    sin_amp_y_ = 0.0;
    sin_freq_x_ = 1.0;
    sin_amp_x_ = 0.0;
    Kp_x_ = 5.0;
    Kd_x_ = 1.0;
    Kp_y_ = 10.0;
    Kd_y_ = 2.0;
    Kp_z_ = 5.0;
    Kd_z_ = 1.0;
  }

#define ARILES_SECTION_ID "BalanceActionParameters"
#define ARILES_CONSTRUCTOR BalanceActionParameters
#define ARILES_ENTRIES                                                                   \
  ARILES_ENTRY_(sin_freq_x)                                                    \
  ARILES_ENTRY_(sin_freq_y)                                                     \
  ARILES_ENTRY_(sin_freq_z)                                                             \
  ARILES_ENTRY_(sin_amp_x)                                                             \
  ARILES_ENTRY_(sin_amp_y)                                                             \
  ARILES_ENTRY_(sin_amp_z)
#include ARILES_INITIALIZE

  double sin_freq_z_, sin_amp_z_;
  double sin_freq_y_, sin_amp_y_;
  double sin_freq_x_, sin_amp_x_;
  double Kp_x_, Kd_x_, Kp_y_, Kd_y_, Kp_z_, Kd_z_;  //com feedback gains
};

class BalanceAction : public WalkingActionBase
{
public:
  BalanceAction() : time_(0.0)
  {
  }

  BalanceAction(ros::NodeHandle &nh, BController *bController);
  virtual ~BalanceAction();

  bool configure(ros::NodeHandle &nh, BController *bController,
                 const property_bag::PropertyBag &parameters) override;

  bool enterHook(const ros::Time &time);

  /// This function is called every cycle of the loop, until the end of the Action()
  bool cycleHook(const ros::Time &time);

  /// Return "true" if the Action has to be stopped. The default implementation use time
  /// to stop the action;
  bool isOverHook(const ros::Time &time);

  /// when isOver()=true, this function is called and the action is removed from the
  /// queue.
  bool endHook(const ros::Time &time);

private:
  eVector2 computePcmp(double K, double omega, const eVector2 &icp_act,
                       const eVector2 &icp_des, const eVector2 &icp_des_vel);

  Eigen::Vector2d computeActICP(double omega, const Eigen::Vector2d &com,
                                const Eigen::Vector2d &com_d,
                                const Eigen::Vector2d &local_coord);

  ros::NodeHandle nh_;
  BController *bc_;
  eVector3 ini_target_;
  ddynamic_reconfigure::DDynamicReconfigurePtr ddr_;
  BalanceActionParameters params_;
  double time_;

  // for real-time subscriber
  ros::Subscriber sub_command_;
  realtime_tools::RealtimeBuffer<std::vector<double>> desired_motions_buffer_;
  std::vector<double> default_motion_command_{0.0, 0.0, 0.0, 0.0, 0.0, 0.0}; // for initialize the realtime buffer
  void commandCallback(const std_msgs::Float64MultiArrayConstPtr& msg);

  std::shared_ptr<realtime_tools::RealtimePublisher<std_msgs::Float64MultiArray>> com_states_pub_;
  double n_com_states_;
  eVector3 current_com_pos_;
};
}

#endif  // BALANCING_ACTION_H


src/static_steps_execution_node.cpp
/*
 * Copyright 2019 PAL Robotics SL. All Rights Reserved
 *
 * Unauthorized copying of this file, via any medium is strictly prohibited,
 * unless it was supplied under the terms of a license agreement or
 * nondisclosure agreement with PAL Robotics SL. In this case it may not be
 * copied or disclosed except in accordance with the terms of that agreement.
 */

#include <pal_locomotion/static_step_executor.h>
#include <math_utils/math_utils.h>
#include <pal_ros_utils/tf_utils.h>
#include <pal_utils/popl.hpp>

using namespace popl;
using namespace pal_locomotion;

BipedStepList createStepList(const eMatrixHom& global_transform, int steps_num,
                             double step_size, double foot_separation,
                             double switch_time = 0.3, double stance_time = 0.9)
{
  ros::Duration switch_duration(switch_time);
  ros::Duration stance_duration(stance_time);

  BipedStepList sl;

  pal_locomotion::Side side;
  for (int i = 1; i < steps_num; ++i)
  {
    double dist_y = foot_separation;
    if (i % 2 == 0)
    {
      side = Side::RIGHT;
      dist_y *= -1;
    }
    else
    {
      side = Side::LEFT;
    }
    eMatrixHom step_pose =
        createMatrix(eVector3(0., 0., 0.), eVector3(step_size * i, dist_y, 0.0));
    Step step(switch_duration, stance_duration, global_transform * step_pose, side);
    ROS_ERROR_STREAM((global_transform * step_pose).translation().transpose());
    sl.push_back(step);
  }
  side = (side._value == Side::RIGHT) ? Side::LEFT : Side::RIGHT;
  foot_separation = (side._value == Side::RIGHT) ? (foot_separation * -1.0) : foot_separation;
  eMatrixHom last_step_pose = createMatrix(
      eVector3(0., 0., 0.), eVector3(step_size * (steps_num - 1), foot_separation, 0.0));
  Step last_step(switch_duration, stance_duration, global_transform * last_step_pose, side);
  sl.push_back(last_step);

  return sl;
}

int main(int argc, char** argv)
{
  double ds_duration;
  double swing_duration;
  double x_step_dist;
  int step_num;
  double foot_separation;

  Switch help_option("h", "help", "produce help message");

  Value<double> ds_duration_option("d", "ds_duration", "ds duration option", 0., &ds_duration);
  Value<double> swing_duration_option("u", "swing_time", "swing leg time option", 0.,
                                      &swing_duration);
  Value<int> step_num_option("s", "step_num", "Number of steps option", 5, &step_num);
  Value<double> x_step_dist_option("x", "x_dist", "Distance in the x direction", 0.15,
                                   &x_step_dist);
  Value<double> foot_separation_option("y", "foot_separation", "Foot separation", 0.1,
                                       &foot_separation);

  OptionParser op("Allowed options");
  op.add(help_option);
  op.add(ds_duration_option);
  op.add(swing_duration_option);
  op.add(step_num_option);
  op.add(x_step_dist_option);
  op.add(foot_separation_option);

  op.parse(argc, argv);

  // print auto-generated help message
  if (help_option.isSet() || !ds_duration_option.isSet() ||
      !swing_duration_option.isSet() || !x_step_dist_option.isSet())
  {
    std::cout << op << "\n";
    return 0;
  }

  if ((foot_separation < 0.1) || (foot_separation > 0.25))
  {
    ROS_ERROR_STREAM("The foot separation provided is dangerous. It should be between [0.1, 0.25]");
  }
  if (fabs(x_step_dist) > 0.3)
  {
    ROS_ERROR_STREAM("The x distance provided is higher than 0.3");
  }

  ros::init(argc, argv, "test");
  ros::NodeHandle nh;
  ros::NodeHandle private_nh("~");

  eMatrixHom global_transform = pal::getTransform("odom", "base_footprint", ros::Duration(3.0));

  BipedStepList step_list =
      createStepList(global_transform, step_num, x_step_dist, foot_separation);

  StaticStepExecutor executor(nh, ros::Duration(ds_duration),
                              ros::Duration(swing_duration), step_list);
  ROS_INFO_STREAM("Will execute motion for " << step_list.size() << " static steps!");
  executor.execute();
}
